### PR TITLE
Fix(jit): support rmsnorm for hidden_size in {64, 128, 256}

### PR DESCRIPTION
**Pull request by Johnsonms**
## Motivation
  jit_rmsnorm silently failed for hidden_size ∈ {64, 128, 256} and hidden_size = 16384 during benchmarking on B200. The root cause:

  - RMSNormKernel only implemented the CTA norm path (hidden_size > 256), so instantiating it for small hidden sizes triggered a static_assert failure ("Hidden size invalid for RMSNorm") at JIT compile time
  - hidden_size = 16384 exceeded the CTA kernel's supported range (≤ 8192), triggering "Unsupported norm configuration"

  Both cases caused nvcc to exit with status 2, flooding stderr with full compilation logs for every failing (hidden_size, batch_size, dtype) combination.

## Modifications
 ` csrc/elementwise/rmsnorm.cuh`
  - Add rmsnorm_warp kernel: uses tile::Memory<Storage>::warp() and apply_norm_warp<kDim>() — one warp (32 threads) per token, no shared memory needed
  - Add RMSNormWarpKernel<kDim, kUsePDL, DType> struct for kDim ∈ {64, 128, 256}, symmetric to the existing RMSNormKernel

  `python/sglang/jit_kernel/norm.py`
  - Add _is_supported_rmsnorm_hidden_size(hidden_size): returns True for warp sizes {64, 128, 256} and CTA sizes (multiples of 256, in range (256, 8192])
  - Add _rmsnorm_kernel_class(hidden_size): returns "RMSNormWarpKernel" or "RMSNormKernel" based on hidden size
  - _jit_rmsnorm_module now routes to the correct kernel class
  - rmsnorm() raises a clean RuntimeError for unsupported hidden sizes (e.g. 16384) instead of falling through to noisy nvcc failures

  `python/sglang/jit_kernel/tests/test_norm_jit.py`
  - Extend RMSNORM_HIDDEN_SIZES to include [64, 128, 256]
  - Add test_rmsnorm_hidden_size_support: unit tests for _is_supported_rmsnorm_hidden_size covering edge cases {0, 64, 128, 256, 512, 8192, 16384}
  - Add test_rmsnorm_kernel_dispatch: verifies correct kernel class selection for all size categories
  - Add test_rmsnorm_rejects_unsupported_hidden_size: verifies clean RuntimeError is raised for {0, 16384}

## Accuracy Tests
  test_rmsnorm_jit validates correctness against flashinfer reference for all supported hidden sizes including the newly added {64, 128, 256}, across bf16/fp16 and batch_size ∈ {1, 19, 99, 989}.

## Benchmarking and Profiling
Measured on B200 (hidden_size=64, batch_size=1, bf16), jit_rmsnorm with the new warp norm kernel is the fastest among all providers:
image

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

### CODE REVIEW

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh [hunk 1]
```diff
@@ -52,6 +52,88 @@ __global__ void rmsnorm_cta(const RMSNormParams __grid_constant__ params) {
...
+
+  PDLWaitPrimary<kUsePDL>();  // wait for primary kernel
+
+  void* output_ptr = nullptr;
+  Storage output_vec;
+  for (uint32_t i = blockIdx.x; i < num_tokens; i += gridDim.x) {
+    const auto input_ptr = pointer::offset<Float>(input, i * input_stride);
+    const auto input_vec = gmem.load(input_ptr);
+    const auto weight_vec = gmem.load(weight_ptr);
+    if (output_ptr != nullptr) {
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh]: What does this mean?

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh [hunk 2]
```diff
@@ -52,6 +52,88 @@ __global__ void rmsnorm_cta(const RMSNormParams __grid_constant__ params) {
...
+  const auto& [input, weight_ptr, output, input_stride, output_stride, num_tokens, eps] = params;
+  const auto gmem = tile::Memory<Storage>::warp();
+
+  PDLWaitPrimary<kUsePDL>();  // wait for primary kernel
+
+  void* output_ptr = nullptr;
+  Storage output_vec;
+  for (uint32_t i = blockIdx.x; i < num_tokens; i += gridDim.x) {
+    const auto input_ptr = pointer::offset<Float>(input, i * input_stride);
+    const auto input_vec = gmem.load(input_ptr);
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh:63]: `input_ptr + threadIdx.x`?

**Review by HydraQYH — COMMENTED** [review]
Have you run unit tests yourself?

File: python/sglang/jit_kernel/tests/test_norm_jit.py [hunk 3]
```diff
@@ -81,5 +83,50 @@ def test_fused_add_rmsnorm_jit(batch_size, hidden_size, dtype):
...
+        (256, True),
+        (512, True),
+        (8192, True),
+        (16384, False),
+    ],
+)
+def test_rmsnorm_hidden_size_support(hidden_size, expected):
+    from sglang.jit_kernel.norm import _is_supported_rmsnorm_hidden_size
+
+    assert _is_supported_rmsnorm_hidden_size(hidden_size) is expected
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/tests/test_norm_jit.py:107]: Are you serious?

File: python/sglang/jit_kernel/tests/test_norm_jit.py [hunk 4]
```diff
@@ -81,5 +83,50 @@ def test_fused_add_rmsnorm_jit(batch_size, hidden_size, dtype):
...
+        (128, "RMSNormWarpKernel"),
+        (256, "RMSNormWarpKernel"),
+        (512, "RMSNormKernel"),
+        (8192, "RMSNormKernel"),
+    ],
+)
+def test_rmsnorm_kernel_dispatch(hidden_size, expected):
+    from sglang.jit_kernel.norm import _rmsnorm_kernel_class
+
+    assert _rmsnorm_kernel_class(hidden_size) == expected
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/tests/test_norm_jit.py:123]: Can this be called a unit test?

**Review by HydraQYH — COMMENTED** [review]
I don't think these unit tests are necessary; tests for these functionalities are already included in the kernel's unit tests.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh [hunk 5]
```diff
@@ -52,6 +52,88 @@ __global__ void rmsnorm_cta(const RMSNormParams __grid_constant__ params) {
...
+    const auto weight_vec = gmem.load(weight_ptr);
+    if (output_ptr != nullptr) {
+      gmem.store(output_ptr, output_vec);
+    }
+    output_ptr = pointer::offset<Float>(output, i * output_stride);
+    output_vec = norm::apply_norm_warp<kDim>(input_vec, weight_vec, eps);
+  }
+  gmem.store(output_ptr, output_vec);
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh]: This `gmem.store(output_ptr, output_vec);` should be inside a for loop, and the `if` statement inside the for loop is meaningless.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh [hunk 6]
```diff
@@ -52,6 +52,88 @@ __global__ void rmsnorm_cta(const RMSNormParams __grid_constant__ params) {
...
+        .input_stride = SI.unwrap(),
+        .output_stride = SO.unwrap(),
+        .num_tokens = num_tokens,
+        .eps = eps,
+    };
+
+    static constexpr uint32_t kNumThreads = device::kWarpThreads;
+    static const uint32_t max_occupancy = runtime::get_blocks_per_sm(kernel, kNumThreads);
+    static const uint32_t kNumSM = runtime::get_sm_count(device.unwrap().device_id);
+    const auto num_blocks = std::min<uint32_t>(num_tokens, max_occupancy * kNumSM);
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh:121]: It appears that unit tests cannot cover this situation: `num_tokens > max_occupancy * kNumSM`.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh (hunk 5 above)
Comment (HydraQYH) [reply python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh]: Perhaps this is correct, as the for loop writes the token processed in the previous for loop each time, and the token processed in the last for loop must be written back separately. However, this makes the code less readable, and I think it would be better to process a token within a single loop.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh (hunk 1 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh]: This code has been refactored to use a simpler sequential pattern — each token is now fully processed (load → compute → store) within a single loop iteration. The output_ptr != nullptr guard and the post-loop gmem.store are no longer needed and have been removed.  [Benchmarking](https://github.com/sgl-project/sglang/pull/20661#discussion_r2971027967) confirmed the sequential pattern is faster than the pipeline approach (up to ~24% at large batch sizes), so there was no reason to keep the added complexity.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh (hunk 2 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh:63]: input_ptr is the base address of the token. The thread-level offset is handled inside gmem.load, which distributes elements across threads in the warp automatically — no need to manually add threadIdx.x here.

File: python/sglang/jit_kernel/tests/test_norm_jit.py (hunk 3 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/tests/test_norm_jit.py:107]: The kernel tests in sgl-kernel/tests/test_norm.py primarily cover the AOT-compiled sgl_kernel.rmsnorm path, and do not exercise the JIT path—so they don’t cover _is_supported_rmsnorm_hidden_size or _rmsnorm_kernel_class.

These tests serve different purposes.

test_rmsnorm_hidden_size_support and test_rmsnorm_kernel_dispatch focus on validating the JIT Python dispatch logic. They run quickly and do not require a GPU, which makes them suitable for CPU-only CI.

test_rmsnorm_rejects_unsupported_hidden_size explicitly verifies the error path for unsupported hidden sizes (e.g., 0 or 16384), which is not covered by the kernel tests.

Given this, it may be beneficial to keep these tests to ensure full coverage across both dispatch logic and error handling paths.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh (hunk 5 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh]: image
  Key takeaways

  1. hidden_size=8192: Sequential is consistently and significantly faster — up to 19% at large batch. This is the most impactful regime (e.g. Llama/Qwen models with hidden_size=8192).
  2. hidden_size 3072/5120: Results are mixed and small (±3–5%), largely within noise margins.
  3. Small batch (≤128): Pipeline wins by ~1–3% — negligible in practice since absolute latency is already ~4µs.
  4. The pipeline pattern adds register pressure (keeping output_ptr/output_vec live across the loop + a conditional branch), which clearly hurts occupancy at hidden_size=8192.

So replace the pipeline kernels with the sequential variants. The sequential pattern is simpler, more readable, and strictly better for the most common large-model configurations.

File: python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh (hunk 6 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/csrc/elementwise/rmsnorm.cuh:121]: Changed. Added bs=8192

**Comment by Johnsonms** [comment]
> Please provide complete unit test results (screenshots or logs) once the above comments are resolved.

`/sgl-workspace/sglang/python/sglang/jit_kernel# python tests/test_norm_jit.py`

image

File: python/sglang/jit_kernel/tests/test_norm_jit.py [hunk 7]
```diff
@@ -81,5 +89,50 @@ def test_fused_add_rmsnorm_jit(batch_size, hidden_size, dtype):
     torch.testing.assert_close(r_jit, r_ref, rtol=1e-2, atol=1e-2)
 
 
+@pytest.mark.parametrize(
```
Comment (HydraQYH) [inline python/sglang/jit_kernel/tests/test_norm_jit.py:92]: There's only one question: are these new unit tests really necessary?

**Review by HydraQYH — APPROVED** [review]
It seems the kernel implementation is OK. And I don't think some unit tests are necessary, but adding them won't hurt. If you insist they are necessary, you can keep them.

File: python/sglang/jit_kernel/tests/test_norm_jit.py (hunk 7 above)
Comment (Johnsonms) [reply python/sglang/jit_kernel/tests/test_norm_jit.py:92]: I will remove that
